File: daybreak/src/daybreak/core/theme_transform.py
from daybreak.colors import adjust_color_for_contrast


def _fallback(value, default):
    return value if value else default


def palette_to_tokens(palette: dict, mode: str) -> dict:
    special = palette.get("special", {})
    colors = palette.get("colors", {})

    bg = _fallback(special.get("background"), "#111111" if mode == "dark" else "#f5f5f5")
    fg = _fallback(special.get("foreground"), "#e5e5e5" if mode == "dark" else "#121212")
    cursor = _fallback(special.get("cursor"), fg)

    muted = _fallback(colors.get("8"), fg)
    primary = _fallback(colors.get("4"), fg)
    success = _fallback(colors.get("2"), fg)
    warning = _fallback(colors.get("3"), fg)
    error = _fallback(colors.get("1"), fg)
    info = _fallback(colors.get("6"), primary)

    surface_1 = _fallback(colors.get("18"), bg)
    surface_2 = _fallback(colors.get("19"), surface_1)
    surface_3 = _fallback(colors.get("20"), muted)

    tokens = {
        "bg": bg,
        "fg": fg,
        "cursor": cursor,
        "muted": muted,
        "primary": primary,
        "success": success,
        "warning": warning,
        "error": error,
        "info": info,
        "surface_1": surface_1,
        "surface_2": surface_2,
        "surface_3": surface_3,
    }

    # Keep semantic foreground tokens readable against the main surface.
    for key in ("fg", "muted", "primary", "success", "warning", "error", "info"):
        tokens[key] = adjust_color_for_contrast(tokens[key], bg, min_ratio=4.0)

    return tokens


def palette_to_accent_tokens(palette: dict, mode: str) -> dict:
    """Derive semantic accent tokens from a palette using contrast-safe utilities."""
    special = palette.get("special", {})
    colors = palette.get("colors", {})

    bg = _fallback(special.get("background"), "#111111" if mode == "dark" else "#f5f5f5")

    accent_primary = _fallback(colors.get("4"), _fallback(special.get("foreground"), bg))
    accent_secondary = _fallback(colors.get("5"), accent_primary)
    accent_success = _fallback(colors.get("2"), _fallback(special.get("foreground"), bg))
    accent_warning = _fallback(colors.get("3"), accent_primary)
    accent_error = _fallback(colors.get("1"), accent_primary)
    accent_selection = _fallback(colors.get("18"), bg)

    accents = {
        "accent_primary": accent_primary,
        "accent_secondary": accent_secondary,
        "accent_success": accent_success,
        "accent_warning": accent_warning,
        "accent_error": accent_error,
        "accent_selection": accent_selection,
    }

    # Ensure foreground accent colors meet a minimum contrast against background.
    for key in ("accent_primary", "accent_secondary", "accent_success", "accent_warning", "accent_error"):
        accents[key] = adjust_color_for_contrast(accents[key], bg, min_ratio=3.0)

    return accents

File: src/core/themeTransform.ts
import { adjustColorForContrast, Palette } from '../colors';

function fallback(value: string | undefined, def: string): string {
  return value || def;
}

export function paletteToTokens(palette: Palette, mode: string): Record<string, string> {
  const special = palette.special || { background: '', foreground: '', cursor: '' };
  const colors = palette.colors || {};

  const bg = fallback(special.background, mode === 'dark' ? '#111111' : '#f5f5f5');
  const fg = fallback(special.foreground, mode === 'dark' ? '#e5e5e5' : '#121212');
  const cursor = fallback(special.cursor, fg);

  const muted = fallback(colors['8'], fg);
  const primary = fallback(colors['4'], fg);
  const success = fallback(colors['2'], fg);
  const warning = fallback(colors['3'], fg);
  const error = fallback(colors['1'], fg);
  const info = fallback(colors['6'], primary);

  const surface1 = fallback(colors['18'], bg);
  const surface2 = fallback(colors['19'], surface1);
  const surface3 = fallback(colors['20'], muted);

  const tokens: Record<string, string> = {
    bg,
    fg,
    cursor,
    muted,
    primary,
    success,
    warning,
    error,
    info,
    surface_1: surface1,
    surface_2: surface2,
    surface_3: surface3,
  };

  for (const key of ['fg', 'muted', 'primary', 'success', 'warning', 'error', 'info']) {
    tokens[key] = adjustColorForContrast(tokens[key], bg, 4.0);
  }

  return tokens;
}

export function paletteToAccentTokens(palette: Palette, mode: string): Record<string, string> {
  const special = palette.special || { background: '', foreground: '', cursor: '' };
  const colors = palette.colors || {};

  const bg = fallback(special.background, mode === 'dark' ? '#111111' : '#f5f5f5');

  const accentPrimary = fallback(colors['4'], fallback(special.foreground, bg));
  const accentSecondary = fallback(colors['5'], accentPrimary);
  const accentSuccess = fallback(colors['2'], fallback(special.foreground, bg));
  const accentWarning = fallback(colors['3'], accentPrimary);
  const accentError = fallback(colors['1'], accentPrimary);
  const accentSelection = fallback(colors['18'], bg);

  const accents: Record<string, string> = {
    accent_primary: accentPrimary,
    accent_secondary: accentSecondary,
    accent_success: accentSuccess,
    accent_warning: accentWarning,
    accent_error: accentError,
    accent_selection: accentSelection,
  };

  for (const key of ['accent_primary', 'accent_secondary', 'accent_success', 'accent_warning', 'accent_error']) {
    accents[key] = adjustColorForContrast(accents[key], bg, 3.0);
  }

  return accents;
}
